cklib/cklib/x509.py

```diff
@@ -0,0 +1,228 @@
+            ),
+            critical=True,
+        )
+        .sign(ca_key, hashes.SHA256(), default_backend())
+    )
+    return ca_key, ca_cert
+
+
+def gen_csr(
+    csr_key: RSAPrivateKey,
+    common_name: str = "some.engineering",
+    san_dns_names: Optional[List[str]] = None,
+    san_ip_addresses: Optional[List[str]] = None,
+) -> CertificateSigningRequest:
+    if san_dns_names is None:
+        san_dns_names = []
+    if san_ip_addresses is None:
+        san_ip_addresses = []
+    csr_build = x509.CertificateSigningRequestBuilder().subject_name(
+        x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, common_name)])
+    )
+    if len(san_dns_names) + len(san_ip_addresses) > 0:
+        csr_build = csr_build.add_extension(
+            x509.SubjectAlternativeName(
+                [x509.DNSName(n) for n in san_dns_names]
+                + [x509.IPAddress(make_ip(i)) for i in san_ip_addresses]
+            ),
+            critical=False,
+        )
+    return csr_build.sign(csr_key, hashes.SHA256(), default_backend())
+
+
+def sign_csr(
+    csr: CertificateSigningRequest,
+    ca_key: RSAPrivateKey,
+    ca_cert: Certificate,
+    days_valid: int = 365,
+    server_auth: bool = True,
+    client_auth: bool = True,
+) -> Certificate:
+    crt_build = (
+        x509.CertificateBuilder()
+        .subject_name(csr.subject)
+        .issuer_name(ca_cert.subject)
+        .public_key(csr.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.utcnow().replace(tzinfo=timezone.utc))
+        .not_valid_after(
+            datetime.utcnow().replace(tzinfo=timezone.utc) + timedelta(days=days_valid)
+        )
+        .add_extension(
+            x509.KeyUsage(
+                digital_signature=True,
+                key_encipherment=True,
+                key_cert_sign=False,
+                key_agreement=False,
+                content_commitment=False,
+                data_encipherment=False,
+                crl_sign=False,
+                encipher_only=False,
+                decipher_only=False,
+            ),
+            critical=True,
+        )
+    )
+    if server_auth or client_auth:
+        key_usage = []
+        if server_auth:
+            key_usage.append(x509.ExtendedKeyUsageOID.SERVER_AUTH)
+        if client_auth:
+            key_usage.append(x509.ExtendedKeyUsageOID.CLIENT_AUTH)
+        crt_build = crt_build.add_extension(
+            x509.ExtendedKeyUsage(key_usage), critical=False
+        )
+    for extension in csr.extensions:
+        if not isinstance(extension.value, x509.SubjectAlternativeName):
+            continue
+        crt_build = crt_build.add_extension(
+            extension.value, critical=extension.critical
+        )
+    return crt_build.sign(ca_key, hashes.SHA256(), default_backend())
+
+
+def write_csr_to_file(csr: CertificateSigningRequest, csr_path: str) -> None:
+    return write_cert_to_file(csr, csr_path)
```

Comment:
a signing request is a certificate?

Reply:
Yes, it is a certificate of a special type signed with the private key of the cert requester. The resulting certificate is then signed with the private key of the CA but issued for the public key part of the private key that the request was signed with.